File: macos/Cleer/DSP/Dereverb.swift
import Foundation

/// Late-reverberation suppression by spectral gain — Swift port of
/// `dereverb.py`. Models the reverb tail as a delayed, decayed copy of recent
/// band power and subtracts it.
final class Dereverb {
    struct Params {
        var rt60 = 0.6
        var delayMs = 40.0
        var strength = 1.0
        var gainFloorDB = -15.0
        var smooth = 0.6
    }

    var params: Params
    var enabled = true
    private let bins: Int
    private let delay: Int
    private let decay: Double
    private var powerHist: [[Double]] = []
    private var tail: [Double]

    init(bins: Int, sampleRate: Double, hop: Int, params: Params = Params()) {
        self.bins = bins
        self.params = params
        let frameS = Double(hop) / sampleRate
        self.delay = max(1, Int((params.delayMs / 1000.0 / frameS).rounded()))
        let decayPerFrame = pow(10.0, -3.0 * frameS / params.rt60)
        self.decay = pow(decayPerFrame, Double(self.delay))
        self.tail = [Double](repeating: 0, count: bins)
    }

    func mask(_ mag: [Double]) -> [Double] {
        guard enabled else { return [Double](repeating: 1, count: bins) }
        var power = [Double](repeating: 0, count: bins)
        for k in 0..<bins { power[k] = mag[k] * mag[k] }
        powerHist.append(power)
        if powerHist.count > delay + 1 { powerHist.removeFirst() }

        if powerHist.count > delay {
            let delayed = powerHist[0]
            for k in 0..<bins {
                let est = decay * delayed[k]
                tail[k] = params.smooth * tail[k] + (1 - params.smooth) * est
            }
        }

        let floorGain = pow(10.0, params.gainFloorDB / 20.0)
        var gain = [Double](repeating: 0, count: bins)
        for k in 0..<bins {
            let late = params.strength * tail[k]
            var g = max(power[k] - late, 0.0) / (power[k] + 1e-12)
            g = max(g, floorGain)
            gain[k] = g
        }
        return gain
    }
}

File: reference/defeedback/dereverb.py
"""Late-reverberation suppression by spectral gain.

Reverb is the room smearing each sound across time: a direct-path arrival
followed by an exponentially decaying tail of reflections. The tail at the
current frame is well modelled as a *delayed, decayed copy of the recent
signal power* in each band (this is the statistical model behind Lebart's
spectral-subtraction dereverb and the WPE family).

So we:
  1.  Estimate late-reverb power per bin = decay * (power D frames ago,
      smoothed).
  2.  Subtract it with a spectral gain, floored to avoid artefacts.

Like the denoiser this outputs a per-bin gain in [0,1], so the two masks
simply multiply together before a single inverse STFT.
"""

from __future__ import annotations

import numpy as np


class Dereverb:
    def __init__(
        self,
        sr: float,
        win: int,
        hop: int,
        rt60: float = 0.6,            # assumed room reverberation time (s)
        delay_ms: float = 40.0,       # boundary between "early" (kept) and "late" (removed)
        strength: float = 1.0,        # 0 = off, 1 = full late-tail subtraction
        gain_floor_db: float = -15.0,
        smooth: float = 0.6,          # power smoothing of the tail estimate
    ):
        self.sr = sr
        self.win = win
        self.hop = hop
        self.strength = strength
        self.gain_floor = 10.0 ** (gain_floor_db / 20.0)
        self.smooth = smooth

        frame_s = hop / sr
        self.delay = max(1, int(round((delay_ms / 1000.0) / frame_s)))
        # Per-frame energy decay implied by RT60 (-60 dB over rt60 seconds),
        # advanced by `delay` frames to predict the tail at the current frame.
        decay_per_frame = 10.0 ** (-3.0 * frame_s / rt60)  # -60dB == factor 1e-3
        self.decay = decay_per_frame ** self.delay

        bins = win // 2 + 1
        self.power_hist: list[np.ndarray] = []
        self.tail = np.zeros(bins)

    def mask(self, mag: np.ndarray) -> np.ndarray:
        power = mag ** 2
        self.power_hist.append(power)
        if len(self.power_hist) > self.delay + 1:
            self.power_hist.pop(0)

        if len(self.power_hist) > self.delay:
            delayed = self.power_hist[0]
            est = self.decay * delayed
            self.tail = self.smooth * self.tail + (1 - self.smooth) * est
        late = self.strength * self.tail

        gain = np.maximum(power - late, 0.0) / (power + 1e-12)
        gain = np.maximum(gain, self.gain_floor)
        return gain

    def process_spectrum(self, spec: np.ndarray) -> np.ndarray:
        out = np.empty_like(spec)
        for i in range(spec.shape[0]):
            g = self.mask(np.abs(spec[i]))
            out[i] = spec[i] * g
        return out
